File: kmm/server/src/main/kotlin/com/danioliveira/taskmanager/routes/ProjectMemberRoutes.kt
package com.danioliveira.taskmanager.routes

import com.danioliveira.taskmanager.api.request.ProjectAssignRequest
import com.danioliveira.taskmanager.api.routes.Projects
import com.danioliveira.taskmanager.domain.service.ProjectService
import com.danioliveira.taskmanager.utils.toUuid
import io.ktor.http.HttpStatusCode
import io.ktor.server.auth.authenticate
import io.ktor.server.request.receive
import io.ktor.server.resources.delete
import io.ktor.server.resources.get
import io.ktor.server.resources.post
import io.ktor.server.response.respond
import io.ktor.server.routing.Route
import org.koin.ktor.ext.inject
import kotlin.uuid.ExperimentalUuidApi

/**
 * Routes for project member management following the pattern:
 * GET /v1/projects/{projectId}/users - Get all members in a project
 * POST /v1/projects/{projectId}/assign - Add member to project
 * DELETE /v1/projects/{projectId}/assign/{userId} - Remove member from project
 */
@OptIn(ExperimentalUuidApi::class)
fun Route.projectMemberRoutes() {
    val projectService by inject<ProjectService>()

    authenticate("auth-jwt") {
        // Get users by project: GET /v1/projects/{projectId}/users
        get<Projects.Id.Users> { res ->
            val users = projectService.getUsersByProject(res.parent.projectId.toUuid())
            call.respond(users)
        }

        // Assign user to project: POST /v1/projects/{projectId}/assign
        post<Projects.Id.Assign> { res ->
            val request = call.receive<ProjectAssignRequest>()
            val assignment = projectService.assignUserToProject(
                projectId = res.parent.projectId.toUuid(), 
                userId = request.userId.toUuid(),
                creatorId = userPrincipal()
            )
            call.respond(HttpStatusCode.Created, assignment)
        }

        // Remove user from project: DELETE /v1/projects/{projectId}/assign/{userId}
        delete<Projects.Id.AssignUser> { res ->
            val removed = projectService.removeUserFromProject(
                projectId = res.parent.projectId.toUuid(),
                userId = res.userId.toUuid(),
                creatorId = userPrincipal()
            )
            if (removed) {
                call.respond(HttpStatusCode.NoContent)
            } else {
                call.respond(HttpStatusCode.NotFound)
            }
        }
    }
}

File: task_manager_dart/server/lib/src/routes/project_member_routes.dart
import 'dart:convert';
import 'package:shelf/shelf.dart';
import 'package:shelf_router/shelf_router.dart';
import '../middleware/auth_middleware.dart';
import '../services/project_service.dart';
import '../util/shelf_helpers.dart';
import '../exceptions/custom_exceptions.dart';

/// Routes for project member management following the pattern:
/// GET /<projectId>/users - Get all members in a project
/// POST /<projectId>/assign - Add member to project
/// DELETE /<projectId>/assign/<userId> - Remove member from project
class ProjectMemberRoutes {
  final ProjectServiceImpl _projectService;
  final AuthMiddleware _authMiddleware;
  late final Router router;

  ProjectMemberRoutes(this._projectService, this._authMiddleware) {
    final baseRouter = Router()
      ..get('/<projectId>/users', _getUsersByProject)
      ..post('/<projectId>/assign', _assignUserToProject)
      ..delete('/<projectId>/assign/<userId>', _removeUserFromProject);
    
    // Wrap the router with auth middleware using Pipeline
    final handler = Pipeline()
        .addMiddleware(_authMiddleware.middleware())
        .addHandler(baseRouter.call);
    
    // Create a new router that delegates to the pipeline
    router = Router()
      ..mount('/', handler);
  }

  Future<Response> _getUsersByProject(Request request) async {
    try {
      final projectId = request.params['projectId'];
      if (projectId == null || projectId.isEmpty) {
        throw ValidationException(message: 'Project ID is required');
      }

      final users = await _projectService.getUsersByProject(projectId);
      return okJsonResponse(users.map((u) => u.toJson()).toList());
    } catch (e) {
      rethrow; // Let error handling middleware handle all exceptions
    }
  }

  Future<Response> _assignUserToProject(Request request) async {
    try {
      final creatorId = request.context['userId'] as String;
      final projectId = request.params['projectId'];
      if (projectId == null || projectId.isEmpty) {
        throw ValidationException(message: 'Project ID is required');
      }

      final requestBody = await request.readJsonBody();
      final userId = requestBody['userId'] as String?;
      if (userId == null || userId.isEmpty) {
        throw ValidationException(message: 'User ID is required in request body');
      }

      final assignment = await _projectService.assignUserToProject(
        projectId,
        userId,
        creatorId,
      );

      return Response(
        201,
        body: jsonEncode(assignment),
        headers: {'content-type': 'application/json'},
      );
    } catch (e) {
      rethrow; // Let error handling middleware handle all exceptions
    }
  }

  Future<Response> _removeUserFromProject(Request request) async {
    try {
      final creatorId = request.context['userId'] as String;
      final projectId = request.params['projectId'];
      final userId = request.params['userId'];

      if (projectId == null || projectId.isEmpty) {
        throw ValidationException(message: 'Project ID is required');
      }
      if (userId == null || userId.isEmpty) {
        throw ValidationException(message: 'User ID is required');
      }

      final removed = await _projectService.removeUserFromProject(
        projectId,
        userId,
        creatorId,
      );

      if (removed) {
        return Response(204); // No Content
      } else {
        throw NotFoundException(message: 'Assignment not found');
      }
    } catch (e) {
      rethrow; // Let error handling middleware handle all exceptions
    }
  }
}

